File: task_3/cpp/3(2).cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <string>

using namespace std;

const double EPS = 1e-3;
const int MAX_ITER = 15;

// Проверка диагонального преобладания
bool checkDiagonalDominance(const vector<vector<double>>& A) {
    int n = A.size();
    for (int i = 0; i < n; ++i) {
        double diag = fabs(A[i][i]);
        double sum = 0;
        for (int j = 0; j < n; ++j) {
            if (i != j) sum += fabs(A[i][j]);
        }
        if (diag <= sum) return false;
    }
    return true;
}

// Проверка положительной определённости (критерий Сильвестра)
bool isPositiveDefinite(const vector<vector<double>>& A) {
    int n = A.size();
    for (int k = 1; k <= n; ++k) {
        double det = 1.0;
        for (int i = 0; i < k; ++i) {
            det *= A[i][i];
        }
        if (det <= 0) return false;
    }
    return true;
}

// Печать таблицы итераций
void printSolutionTable(const vector<vector<double>>& history) {
    int n = history[0].size();
    cout << fixed << setprecision(5);

    // Верхняя граница
    cout << "\n╔";
    for (int i = 0; i < n + 2; ++i)
        cout << "════════════════╦";
    cout << "\b╗\n";

    // Заголовок
    cout << "║" << setw(16) << left << "       N";
    for (int i = 0; i < n; ++i) {
        cout << "║" << setw(16) << left << ("       x" + to_string(i + 1));
    }
    cout << "║" << setw(16) << left << "       εn" << " ║\n";

    // Разделитель
    cout << "╠";
    for (int i = 0; i < n + 2; ++i)
        cout << "════════════════╬";
    cout << "\b╣\n";

    // Данные
    for (size_t step = 0; step < history.size(); ++step) {
        cout << "║" << setw(16) << left << ("       " + to_string(step + 1));
        double max_eps = 0;
        for (int i = 0; i < n; ++i) {
            cout << "║   " << setw(13) << left << history[step][i];
            if (step > 0) {
                max_eps = max(max_eps, fabs(history[step][i] - history[step - 1][i]));
            }
        }
        if (step == 0)
            cout << "║" << setw(16) << left << "       -";
        else
            cout << "║" << setw(16) << left << max_eps;
        cout << "║\n";
    }

    // Нижняя граница
    cout << "╚";
    for (int i = 0; i < n + 2; ++i)
        cout << "════════════════╩";
    cout << "\b╝\n";
}

// Преобразование к каноническому виду: x = Cx + f
void transformToCanonical(const vector<vector<double>>& A, const vector<double>& b,
                          vector<vector<double>>& C, vector<double>& f) {
    int n = A.size();
    C = A;
    f = b;
    for (int i = 0; i < n; ++i) {
        double diag = A[i][i];
        f[i] = b[i] / diag;
        for (int j = 0; j < n; ++j) {
            if (i == j)
                C[i][j] = 0;
            else
                C[i][j] = -A[i][j] / diag;
        }
    }
}

// Вывод матрицы C и вектора f
void printCanonicalForm(const vector<vector<double>>& C, const vector<double>& f) {
    cout << "\nМатрица C (коэффициенты итераций):\n";
    for (const auto& row : C) {
        for (double val : row)
            cout << setw(12) << fixed << setprecision(6) << val;
        cout << "\n";
    }
    cout << "\nВектор f (свободные члены):\n";
    for (double val : f)
        cout << setw(12) << fixed << setprecision(6) << val << "\n";
}

// Метод Зейделя
vector<double> seidelMethod(const vector<vector<double>>& A, const vector<double>& b,
                            vector<vector<double>>& history, int& iterations_out) {
    int n = A.size();
    vector<double> x(n, 0.0), prev_x(n, 0.0);

    for (int iter = 0; iter < MAX_ITER; ++iter) {
        for (int i = 0; i < n; ++i) {
            double sum = b[i];
            for (int j = 0; j < n; ++j) {
                if (j < i)
                    sum -= A[i][j] * x[j];
                else if (j > i)
                    sum -= A[i][j] * prev_x[j];
            }
            x[i] = sum / A[i][i];
        }

        history.push_back(x);

        double max_delta = 0;
        for (int i = 0; i < n; ++i) {
            max_delta = max(max_delta, fabs(x[i] - prev_x[i]));
        }

        if (max_delta < EPS) {
            iterations_out = iter + 1;
            break;
        }

        prev_x = x;
    }

    return x;
}

int main() {
    vector<vector<double>> A = {
        {0.87, -0.04, 0.21, -18.0},
        {0.25, -1.23, -0.19, -0.09},
        {-0.21, -0.19, 0.80, -0.13},
        {0.15, -1.31, 0.06, 1.08}
    };

    vector<double> b = {-1.24, 1.08, 2.56, 0.87};

    // Проверка условий
    cout << (checkDiagonalDominance(A) ? "Выполняется диагональное преобладание.\n"
                                       : "Предупреждение: диагональное преобладание не выполняется. Метод может не сойтись.\n");
    cout << (isPositiveDefinite(A) ? "Матрица A положительно определённая.\n"
                                   : "Матрица A не является положительно определённой.\n");

    // Преобразование
    vector<vector<double>> C;
    vector<double> f;
    transformToCanonical(A, b, C, f);
    printCanonicalForm(C, f);

    // Итерации
    vector<vector<double>> history;
    int iterations = 0;
    vector<double> solution = seidelMethod(A, b, history, iterations);

    // Вывод
    cout << "\nРешение системы (метод Зейделя):\n";
    for (int i = 0; i < solution.size(); ++i)
        cout << "x" << (i + 1) << " = " << fixed << setprecision(6) << solution[i] << endl;

    cout << "\nТаблица итераций:\n";
    printSolutionTable(history);

    // Расчёт количества операций
    int n = A.size();
    int operations = 2 * n * n * iterations;
    cout << "\nКоличество операций (по формуле 2n^2 * k): " << operations << endl;

    return 0;
}

File: task_3/go/3(2).go
package main

import (
	"fmt"
	"math"
	"strconv"
)

const (
	EPS      = 1e-3
	MAX_ITER = 15
)

// Проверка диагонального преобладания
func checkDiagonalDominance(A [][]float64) bool {
	n := len(A)
	for i := 0; i < n; i++ {
		diag := math.Abs(A[i][i])
		sum := 0.0
		for j := 0; j < n; j++ {
			if i != j {
				sum += math.Abs(A[i][j])
			}
		}
		if diag <= sum {
			return false
		}
	}
	return true
}

// Проверка положительной определённости (упрощённый критерий Сильвестра)
func isPositiveDefinite(A [][]float64) bool {
	n := len(A)
	for k := 1; k <= n; k++ {
		det := 1.0
		for i := 0; i < k; i++ {
			det *= A[i][i]
		}
		if det <= 0 {
			return false
		}
	}
	return true
}

// Печать таблицы итераций
func printSolutionTable(history [][]float64) {
	n := len(history[0])
	fmt.Println()

	// Верхняя граница
	fmt.Print("╔")
	for i := 0; i < n+2; i++ {
		fmt.Print("════════════════╦")
	}
	fmt.Print("\b╗\n")

	// Заголовок
	fmt.Printf("║%-16s", "       N")
	for i := 0; i < n; i++ {
		fmt.Printf("║%-16s", "       x"+strconv.Itoa(i+1))
	}
	fmt.Printf("║%-15s ║\n", "       εn")

	// Разделитель
	fmt.Print("╠")
	for i := 0; i < n+2; i++ {
		fmt.Print("════════════════╬")
	}
	fmt.Print("\b╣\n")

	// Данные
	for step := 0; step < len(history); step++ {
		fmt.Printf("║%-16s", "       "+strconv.Itoa(step+1))
		maxEps := 0.0
		for i := 0; i < n; i++ {
			fmt.Printf("║   %-13.5f", history[step][i])
			if step > 0 {
				maxEps = math.Max(maxEps, math.Abs(history[step][i]-history[step-1][i]))
			}
		}
		if step == 0 {
			fmt.Printf("║%-16s", "       -")
		} else {
			fmt.Printf("║%-16.5f", maxEps)
		}
		fmt.Println("║")
	}

	// Нижняя граница
	fmt.Print("╚")
	for i := 0; i < n+2; i++ {
		fmt.Print("════════════════╩")
	}
	fmt.Print("\b╝\n")
}

// Преобразование к каноническому виду: x = Cx + f
func transformToCanonical(A [][]float64, b []float64) (C [][]float64, f []float64) {
	n := len(A)
	C = make([][]float64, n)
	f = make([]float64, n)

	for i := 0; i < n; i++ {
		C[i] = make([]float64, n)
		diag := A[i][i]
		f[i] = b[i] / diag
		for j := 0; j < n; j++ {
			if i == j {
				C[i][j] = 0
			} else {
				C[i][j] = -A[i][j] / diag
			}
		}
	}
	return C, f
}

// Вывод матрицы C и вектора f
func printCanonicalForm(C [][]float64, f []float64) {
	fmt.Println("\nМатрица C (коэффициенты итераций):")
	for _, row := range C {
		for _, val := range row {
			fmt.Printf("%12.6f", val)
		}
		fmt.Println()
	}

	fmt.Println("\nВектор f (свободные члены):")
	for _, val := range f {
		fmt.Printf("%12.6f\n", val)
	}
}

// Метод Зейделя
func seidelMethod(A [][]float64, b []float64, history *[][]float64) ([]float64, int) {
	n := len(A)
	x := make([]float64, n)
	prevX := make([]float64, n)
	iterations := 0

	for iter := 0; iter < MAX_ITER; iter++ {
		for i := 0; i < n; i++ {
			sum := b[i]
			for j := 0; j < n; j++ {
				if j < i {
					sum -= A[i][j] * x[j]
				} else if j > i {
					sum -= A[i][j] * prevX[j]
				}
			}
			x[i] = sum / A[i][i]
		}

		// Добавляем копию текущего решения в историю
		xCopy := make([]float64, n)
		copy(xCopy, x)
		*history = append(*history, xCopy)

		// Проверка сходимости
		maxDelta := 0.0
		for i := 0; i < n; i++ {
			maxDelta = math.Max(maxDelta, math.Abs(x[i]-prevX[i]))
		}

		if maxDelta < EPS {
			iterations = iter + 1
			break
		}

		copy(prevX, x)
	}

	return x, iterations
}

func main() {
	A := [][]float64{
		{0.87, -0.04, 0.21, -18.0},
		{0.25, -1.23, -0.19, -0.09},
		{-0.21, -0.19, 0.80, -0.13},
		{0.15, -1.31, 0.06, 1.08},
	}

	b := []float64{-1.24, 1.08, 2.56, 0.87}

	// Проверка условий
	if checkDiagonalDominance(A) {
		fmt.Println("Выполняется диагональное преобладание.")
	} else {
		fmt.Println("Предупреждение: диагональное преобладание не выполняется. Метод может не сойтись.")
	}

	if isPositiveDefinite(A) {
		fmt.Println("Матрица A положительно определённая.")
	} else {
		fmt.Println("Матрица A не является положительно определённой.")
	}

	// Преобразование
	C, f := transformToCanonical(A, b)
	printCanonicalForm(C, f)

	// Итерации
	var history [][]float64
	solution, iterations := seidelMethod(A, b, &history)

	// Вывод
	fmt.Println("\nРешение системы (метод Зейделя):")
	for i := 0; i < len(solution); i++ {
		fmt.Printf("x%d = %.6f\n", i+1, solution[i])
	}

	fmt.Println("\nТаблица итераций:")
	printSolutionTable(history)

	// Расчёт количества операций
	n := len(A)
	if iterations > 0 {
		operations := 2 * n * n * iterations
		fmt.Printf("\nКоличество операций (по формуле 2n^2 * k): %d\n", operations)
	} else {
		fmt.Println("\nМетод не сошёлся — количество операций не рассчитывается.")
	}
}
